File: Day6-TodoList/app/src/main/java/com/byted/camp/todolist/ui/NoteViewHolder.java
package com.byted.camp.todolist.ui;

import android.graphics.Color;
import android.graphics.Paint;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import android.util.Log;
import android.view.View;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.TextView;
import android.widget.Toast;

import com.byted.camp.todolist.MainActivity;
import com.byted.camp.todolist.NoteOperator;
import com.byted.camp.todolist.R;
import com.byted.camp.todolist.TodoListViewModel;
import com.byted.camp.todolist.TodoRepository;
import com.byted.camp.todolist.beans.Note;
import com.byted.camp.todolist.beans.State;
import com.byted.camp.todolist.db.room.TodoEntity;

import java.text.SimpleDateFormat;
import java.util.Locale;

import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Action;

/**
 * Created on 2019/1/23.
 *
 * @author xuyingyi@bytedance.com (Yingyi Xu)
 */
public class NoteViewHolder extends RecyclerView.ViewHolder {

    private static final SimpleDateFormat SIMPLE_DATE_FORMAT =
            new SimpleDateFormat("EEE, d MMM yyyy HH:mm:ss", Locale.ENGLISH);

    private final TodoListViewModel viewModel;

    private CheckBox checkBox;
    private TextView contentText;
    private TextView dateText;
    private View deleteBtn;

    public NoteViewHolder(@NonNull View itemView, TodoListViewModel viewModel) {
        super(itemView);
        this.viewModel = viewModel;

        checkBox = itemView.findViewById(R.id.checkbox);
        contentText = itemView.findViewById(R.id.text_content);
        dateText = itemView.findViewById(R.id.text_date);
        deleteBtn = itemView.findViewById(R.id.btn_delete);
    }

    public void bind(final TodoEntity note) {
        contentText.setText(note.getContent());
        dateText.setText(SIMPLE_DATE_FORMAT.format(note.getDate()));

        checkBox.setOnCheckedChangeListener(null);
        checkBox.setChecked(note.getState() == State.DONE);
        checkBox.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                note.setState(isChecked ? State.DONE : State.TODO);
                Disposable d = viewModel.update(note).subscribe(new Action() {
                    @Override
                    public void run() {
                        Log.d("[BIND_LIVE]", "update success");
                    }
                });
            }
        });
        deleteBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Disposable d = viewModel.delete(note).subscribe(new Action() {
                    @Override
                    public void run() throws Exception {
                        Log.d("[BIND_LIVE]", "delete success");
                    }
                });
            }
        });

        if (note.getState() == State.DONE) {
            contentText.setTextColor(Color.DKGRAY);
            contentText.setPaintFlags(contentText.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);
        } else {
            contentText.setTextColor(Color.BLACK);
            contentText.setPaintFlags(contentText.getPaintFlags() & ~Paint.STRIKE_THRU_TEXT_FLAG);
        }

        itemView.setBackgroundColor(note.getPriority().color);
    }
}

File: Day6-TodoList/app/src/main/java/com/byted/camp/todolist/viewController/adapters/NoteViewHolder.kt
package com.byted.camp.todolist.ViewController.Adapters

import android.graphics.Color
import android.graphics.Paint
import androidx.recyclerview.widget.RecyclerView

import android.util.Log
import android.view.View
import android.widget.CheckBox
import android.widget.CompoundButton
import android.widget.TextView

import com.byted.camp.todolist.R
import com.byted.camp.todolist.ViewModel.TodoListViewModel
import com.byted.camp.todolist.db.beans.State
import com.byted.camp.todolist.db.room.TodoEntity

import java.text.SimpleDateFormat
import java.util.Locale

import io.reactivex.disposables.Disposable
import io.reactivex.functions.Action

/**
 * Created on 2019/1/23.
 *
 * @author xuyingyi@bytedance.com (Yingyi Xu)
 */
class NoteViewHolder(itemView: View, private val viewModel: TodoListViewModel) : RecyclerView.ViewHolder(itemView) {

    private val checkBox: CheckBox
    private val contentText: TextView
    private val dateText: TextView
    private val deleteBtn: View

    init {

        checkBox = itemView.findViewById(R.id.checkbox)
        contentText = itemView.findViewById(R.id.text_content)
        dateText = itemView.findViewById(R.id.text_date)
        deleteBtn = itemView.findViewById(R.id.btn_delete)
    }

    fun bind(note: TodoEntity) {
        contentText.text = note.content
        dateText.text = SIMPLE_DATE_FORMAT.format(note.date)

        checkBox.setOnCheckedChangeListener(null)
        checkBox.isChecked = note.state === State.DONE
        checkBox.setOnCheckedChangeListener { buttonView, isChecked ->
            note.state = if (isChecked) State.DONE else State.TODO
            val d = viewModel.update(note).subscribe { Log.d("[BIND_LIVE]", "update success") }
        }
        deleteBtn.setOnClickListener { val d = viewModel.delete(note).subscribe { Log.d("[BIND_LIVE]", "delete success") } }

        if (note.state === State.DONE) {
            contentText.setTextColor(Color.DKGRAY)
            contentText.paintFlags = contentText.paintFlags or Paint.STRIKE_THRU_TEXT_FLAG
        } else {
            contentText.setTextColor(Color.BLACK)
            contentText.paintFlags = contentText.paintFlags and Paint.STRIKE_THRU_TEXT_FLAG.inv()
        }

        itemView.setBackgroundColor(note.priority!!.color)
    }

    companion object {

        private val SIMPLE_DATE_FORMAT = SimpleDateFormat("EEE, d MMM yyyy HH:mm:ss", Locale.ENGLISH)
    }
}
